Add support for Aquastream Ultimate sensors

aleksamagicka opened the pull request:
This adds support for Aquastream Ultimate sensors, reverse engineered and tested in #50.

leoratte reviewed (changes requested):
Thanks for adding support for the Aquastream Ultimate.
This looks mostly good so far. The only issue I see at the moment is that the pump and fan readings are all implemented as special cases instead of using NUM_FANS.

aleksamagicka commented:
Yes, that's next, I'll have to rearrange the labels to match, since the fan will take up the first slot.

aleksamagicka commented:
I moved the fan to the `num_fans` method, the pump offsets are too all over the place for that. Also added what I think is the pump frequency, but that, as well as flow sensors, will have to be tested to iron out.

leoratte commented on aquacomputer_d5next.c:
```diff
@@ -842,6 +916,11 @@ static umode_t aqc_is_visible(const void *data, enum hwmon_sensor_types type, u3
 	case hwmon_curr:
 		switch (priv->kind) {
 		case aquastreamxt:
+			/* Special case to support pump and fan current */
+			if (channel < 2)
+				return 0444;
+			break;
+		case aquastreamult:
```
You might have mixed up Aquastream XT and Ultimate here.

aleksamagicka replied:
Embarrassing mistakes, thanks for noticing!

leoratte commented on aquacomputer_d5next.c:
```diff
@@ -1958,6 +2066,8 @@ static int aqc_probe(struct hid_device *hdev, const struct hid_device_id *id)
 		priv->secondary_ctrl_report_size = AQUASTREAMXT_SECONDARY_CTRL_REPORT_SIZE;
 		priv->secondary_ctrl_report = aquastreamxt_secondary_ctrl_report;
+	} else if (priv->kind == aquastreamult) {
+		priv->fan_structure = &aqc_aquastreamult_fan_structure;
 	} else {
 		priv->serial_number_start_offset = AQC_SERIAL_START;
 		priv->firmware_version_offset = AQC_FIRMWARE_VERSION;
```
I assume this won't work as the serial number and firmware version offsets are never set for the Aquastream Ultimate.

leoratte reviewed (changes requested):
Thanks for the work. Please see my comments below.

leoratte commented on aquacomputer_d5next.c:
```diff
@@ -458,6 +482,35 @@ static const char *const label_aquastreamxt_temp_sensors[] = {
+	"Pump speed",
+	"Pump frequency",
+	"Pressure [mbar]",
+	"Flow speed 1"
```
Same as above.

leoratte commented on aquacomputer_d5next.c:
```diff
@@ -163,6 +166,27 @@ static u16 d5next_sensor_fan_offsets[] = { D5NEXT_PUMP_OFFSET, D5NEXT_FAN_OFFSET
+#define AQUASTREAMULT_FAN_OFFSET		0x41
+#define AQUASTREAMULT_PUMP_FREQUENCY		0x15
+#define AQUASTREAMULT_PRESSURE_OFFSET		0x57
+#define AQUASTREAMULT_FLOW_1			0x37
```
Rename to 'AQUASTREAMULT_FLOW' as there is only one flow sensors.

leoratte commented on aquacomputer_d5next.c:
```diff
@@ -1609,6 +1689,21 @@ static int aqc_raw_event(struct hid_device *hdev, struct hid_report *report, u8
 		}
 		break;
+	case aquastreamult:
+		/* Second temp sensor is not positioned after the first one, read it here */
+		priv->temp_input[1] = get_unaligned_be16(data + AQUASTREAMULT_EXTERNAL_TEMP) * 10;
```
The addresses you provide above are directly after each other.

leoratte commented on aquacomputer_d5next.c:
```diff
@@ -458,6 +482,35 @@ static const char *const label_aquastreamxt_temp_sensors[] = {
+};
+
+static const char *const label_aquastreamult_voltages[] = {
+	"Fan +12V",
```
The fan can be voltage controlled. so this is not always 12V.

leoratte reviewed (changes requested):
I found some more small mistakes.

leoratte commented on aquacomputer_d5next.c:
```diff
@@ -163,6 +166,27 @@ static u16 d5next_sensor_fan_offsets[] = { D5NEXT_PUMP_OFFSET, D5NEXT_FAN_OFFSET
+
+/* Sensor report offsets for the Aquastream Ultimate pump */
+#define AQUASTREAMULT_COOLANT_TEMP		0x2D
+#define AQUASTREAMULT_EXTERNAL_TEMP		0x2F
```
Remove this as it is not used anymore.

aleksamagicka replied:
Thanks!

leoratte commented:
Should be ready to merge. Just add a short description of the Aquastream Ultimate capabilities to the docs.

aleksamagicka commented:
Yeah, might as well do it here. Definitely need to clean it up a bit later.

leoratte commented:
The aquastreamult_fan_sensor_offsets are missing causing the driver to crash.
